[docs] Add migration guides for Calendar, MediaLibrary, Contacts (#45449)
expo/expo: docs/pages/guides/sdk-libraries-migration/calendar.mdx

---

> title: Migrate to Calendar@Next

Reviewer: ```suggestion
title: Migrate to expo-calendar/next
```

Let's use the library name itself for the following reasons:
- Good for internal search
- Good for SEO

> sidebar_title: Calendar@Next

Reviewer: ```suggestion
sidebar_title: expo-calendar/next
```

> description: Migrate from the legacy expo-calendar API to the new class-based @Next API with ExpoCalendar, ExpoCalendarEvent, and hooks.

Reviewer: ```suggestion
description: Migrate from the legacy expo-calendar API to the new class-based expo-calendar/next API with ExpoCalendar, ExpoCalendarEvent, and hooks.
```

---

> The `expo-calendar/next` API is now stable. The legacy `expo-calendar` module is deprecated. Migrate to `expo-calendar/next` to benefit from the new API and future fixes.

Reviewer: ```suggestion
```

We need an uninstall/install section in these guides. People usually find it hard to navigate between different docs for such a trivial step.

I'd suggest adding an Installation section before the first H2 and using Terminal components on how to uninstall and install the libraries.

Also easier for AI agents to parse the information.

The new API replaces free functions that accepted ids with methods on class instances. Calendars, events, reminders, and attendees are now represented as class instances that hold only the id of the native object. Key changes:

> - Operations on calendars, events, reminders, and attendees are now methods on the corresponding instance instead of free functions that accept an id.

Reviewer: ```suggestion
- Operations on calendars, events, reminders, and attendees are now methods on the corresponding instance instead of free functions that accept an ID.
```

For consistency

> - `createCalendar`, `createEvent`, and `createReminder` return class instances instead of string ids.

Reviewer: ```suggestion
- `createCalendar`, `createEvent`, and `createReminder` return class instances instead of string IDs.
```

## Importing the new API

Import from `expo-calendar/next`:

```ts
import { ExpoCalendar, ExpoCalendarEvent } from 'expo-calendar/next';
```

## Calendars

### Create a calendar

```ts
// Before
const calendarId = await Calendar.createCalendarAsync({ title: 'My Calendar', color: '#ff0000' });

// After
const calendar = await createCalendar({ title: 'My Calendar', color: '#ff0000' });
```

> `createCalendar` returns an `ExpoCalendar` instance, not just an id.

Reviewer: ```suggestion
```ts
// Before
const calendarId = await Calendar.createCalendarAsync({ title: 'My Calendar', color: '#ff0000' });

// After
const calendar = await createCalendar({ title: 'My Calendar', color: '#ff0000' });
```

`createCalendar` returns an `ExpoCalendar` instance, not just an ID.
```

### List calendars

```ts
// Before
const calendars = await Calendar.getCalendarsAsync(Calendar.EntityTypes.EVENT);

// After
const calendars = await getCalendars(EntityTypes.EVENT);
```

### Get a calendar by ID

```ts
// Before
// No direct equivalent, had to filter from getCalendarsAsync

// After
const calendar = await ExpoCalendar.get(calendarId);
```

### Update a calendar

```ts
// Before
await Calendar.updateCalendarAsync(calendarId, { title: 'Renamed' });

// After
await calendar.update({ title: 'Renamed' });
```

### Delete a calendar

```ts
// Before
await Calendar.deleteCalendarAsync(calendarId);

// After
await calendar.delete();
```

### Get default calendar (iOS only)

```ts
// Before
const calendar = await Calendar.getDefaultCalendarAsync();

// After
const calendar = getDefaultCalendarSync();
```

### Show a calendar picker (iOS only)

```ts
// Before
// No equivalent

// After
const calendar = await presentPicker();
if (calendar) {
  // user selected a calendar
}
```

> `presentPicker` returns `null` if the user dismisses the picker without selecting a calendar.

Reviewer: ```suggestion
`presentPicker` returns `null` if the app user dismisses the picker without selecting a calendar.
```

## Events

### Create an event

```ts
// Before
const eventId = await Calendar.createEventAsync(calendarId, {
  title: 'Lunch',
  startDate,
  endDate,
});

// After
const event = await calendar.createEvent({ title: 'Lunch', startDate, endDate });
```

> `createEvent` returns an `ExpoCalendarEvent` instance, not just an id.

Reviewer: ```suggestion
`createEvent` returns an `ExpoCalendarEvent` instance, not just an ID.
```

### List events in a calendar

```ts
// Before
const events = await Calendar.getEventsAsync([calendarId], startDate, endDate);

// After
const events = await calendar.listEvents(startDate, endDate);
```

### List events across multiple calendars

```ts
// Before
const events = await Calendar.getEventsAsync([id1, id2], startDate, endDate);

// After
const events = await listEvents([calendar1, calendar2], startDate, endDate);
```

### Get an event by ID

```ts
// Before
const event = await Calendar.getEventAsync(eventId);

// After
const event = await ExpoCalendarEvent.get(eventId);
```

### Update an event

```ts
// Before
await Calendar.updateEventAsync(eventId, { title: 'Lunch with Alex' });

// After
await event.update({ title: 'Lunch with Alex' });
```

> The legacy `updateEventAsync` accepted `recurringEventOptions` (iOS only) to target a single occurrence or future occurrences of a recurring event. This is not supported in the new API — `update()` always modifies the entire recurring event series.

Reviewer: ```suggestion
The legacy `updateEventAsync` accepted `recurringEventOptions` (iOS only) to target a single occurrence or future occurrences of a recurring event. This is not supported in the new API &mdash; `update()` always modifies the entire recurring event series.
```

### Delete an event

```ts
// Before
await Calendar.deleteEventAsync(eventId);

// After
await event.delete();
```

> The legacy `deleteEventAsync` accepted `recurringEventOptions` (iOS only) to target a single occurrence or future occurrences of a recurring event. This is not supported in the new API — `delete()` always deletes the entire recurring event series.

Reviewer: ```suggestion
The legacy `deleteEventAsync` accepted `recurringEventOptions` (iOS only) to target a single occurrence or future occurrences of a recurring event. This is not supported in the new API &mdash; `delete()` always deletes the entire recurring event series.
```

### Open event in calendar

```ts
// Before
await Calendar.openEventInCalendarAsync(params);

// After
await event.openInCalendar(params);
```

> The `id` field is no longer part of params — it comes from the event instance. Presentation options (`allowsEditing`, `allowsCalendarPreview`, `startNewActivityTask`) are now passed in the same params object instead of as a separate argument.

Reviewer: ```suggestion
The `id` field is no longer part of params &mdash; it comes from the event instance. Presentation options (`allowsEditing`, `allowsCalendarPreview`, `startNewActivityTask`) are now passed in the same params object instead of as a separate argument.
```

### Edit event with native form

```ts
// Before
await Calendar.editEventInCalendarAsync(params);
// or, to create a new event with form
await Calendar.createEventInCalendarAsync({ title, startDate, endDate });

// After
await event.editInCalendar(params);
// or, to create a new event with form
await calendar.addEventWithForm({ title, startDate, endDate });
```

> The `id` field is no longer part of params — it comes from the event instance. Presentation options (`startNewActivityTask`) are now passed in the same params object instead of as a separate argument.

Reviewer: ```suggestion
The `id` field is no longer part of params &mdash; it comes from the event instance. Presentation options (`startNewActivityTask`) are now passed in the same params object instead of as a separate argument.
```

## Breaking semantic changes

- Calendars, events, reminders, and attendees are now class instances. Operations are methods on the instance instead of free functions that accept an id. Use the corresponding `.get(id)` static method to obtain an instance if you only have an id.
- `createCalendar`, `createEvent`, and `createReminder` return class instances instead of string ids.
- The `Async` suffix is dropped. The majority of the library is asynchronous — only synchronous functions use a `Sync` suffix (e.g., `getDefaultCalendarSync`, `getSourcesSync`, `getOccurrenceSync`).
- `getSourcesAsync` is replaced by the synchronous `getSourcesSync`. Fetching a single source by id has no direct equivalent.
- `createEventInCalendarAsync` is renamed to `calendar.addEventWithForm`.
- `openEventInCalendar` (the fire-and-forget sync variant) is removed. Use `event.openInCalendar()` instead.
- Attendee operations are now instance methods: creating an attendee is done via `event.createAttendee()` on an `ExpoCalendarEvent` instance; updating and deleting are methods on the resulting `ExpoCalendarAttendee` instance.

## Reference

> See the full API at [Calendar (next)](/versions/latest/sdk/calendar-next/).

Reviewer: Please use BoxLink here.

await Calendar.deleteReminderAsync(reminderId);

// After
await reminder.delete();
```

## Sources

```ts
// Before
const sources = await Calendar.getSourcesAsync();

// After
const sources = getSourcesSync();
```

`getSourcesAsync` is replaced by the synchronous `getSourcesSync`. Fetching a single source by ID has no direct equivalent in the new API.

## Permissions

```ts
// Before
await Calendar.requestCalendarPermissionsAsync();
await Calendar.getCalendarPermissionsAsync();
await Calendar.requestRemindersPermissionsAsync();
await Calendar.getRemindersPermissionsAsync();

// After
await requestCalendarPermissions();
await getCalendarPermissions();
await requestRemindersPermissions();
await getRemindersPermissions();
```

`useCalendarPermissions` and `useRemindersPermissions` hooks are unchanged.

## Breaking semantic changes

- Calendars, events, reminders, and attendees are now class instances. Operations are methods on the instance instead of free functions that accept an ID. Use the corresponding `.get(id)` static method to obtain an instance if you only have an ID.
- `createCalendar`, `createEvent`, and `createReminder` return class instances instead of string IDs.

> - The `Async` suffix is dropped. The majority of the library is asynchronous &mdash; only synchronous functions use a `Sync` suffix (e.g., `getDefaultCalendarSync`, `getSourcesSync`, `getOccurrenceSync`).

Reviewer: ```suggestion
- The `Async` suffix is dropped. The majority of the library is asynchronous &mdash; only synchronous functions use a `Sync` suffix (for example, `getDefaultCalendarSync`, `getSourcesSync`, `getOccurrenceSync`).
```

This will fix the Vale warning.
